pkg/sql/copy/copy_test: TestCopyFromRandom failed

pkg/sql/copy/copy_test.TestCopyFromRandom [failed](https://tanzanite.cluster.engflow.com/invocations/default/7ca50a36-034a-4cda-9f5d-92c165c7ccee?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [900f8b66a50ca600527856b9bf83ccbbcbef932a](https://github.com/cockroachdb/cockroach/commits/900f8b66a50ca600527856b9bf83ccbbcbef932a):


```
=== RUN   TestCopyFromRandom
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCopyFromRandom3385764067
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
I240412 09:42:50.762640 1 (gostd) rand.go:250  [-] 1  random seed: 5058866758784072023
    copy_in_test.go:256: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Tenant/10/Table/106/1/164/0): "sql txn" meta={id=9f9d04f7 key=/Tenant/10/Table/106/1/0/0 iso=Serializable pri=0.04074592 epo=0 ts=1712914983.020712413,2 min=1712914981.334040366,0 seq=1001} lock=true stat=PENDING rts=1712914983.020712413,2 wto=false gul=1712914981.834040366,0
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestCopyFromRandom3385764067
--- FAIL: TestCopyFromRandom (16.05s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCopyFromRandom.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37804

pkg/sql/copy/copy_test.TestCopyFromRandom [failed](https://tanzanite.cluster.engflow.com/invocations/default/be232a8b-c406-45e5-a00a-fdbe1a5bf38f?testReportRun=3&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [347cdc76d4c5abb2e872f325e944337a46b5883f](https://github.com/cockroachdb/cockroach/commits/347cdc76d4c5abb2e872f325e944337a46b5883f):


```
=== RUN   TestCopyFromRandom
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCopyFromRandom4264826904
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
I240422 10:37:35.857117 1 (gostd) rand.go:250  [-] 1  random seed: 2334256176139673872
    copy_in_test.go:256: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Tenant/10/Table/106/1/800/0): "sql txn" meta={id=34c80747 key=/Tenant/10/Table/106/1/0/0 iso=Serializable pri=0.00041984 epo=0 ts=1713782267.289322050,2 min=1713782265.341974114,0 seq=1001} lock=true stat=PENDING rts=1713782267.289322050,2 wto=false gul=1713782265.841974114,0
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestCopyFromRandom4264826904
--- FAIL: TestCopyFromRandom (15.10s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCopyFromRandom.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/copy/copy_test.TestCopyFromRandom [failed](https://tanzanite.cluster.engflow.com/invocations/default/ca69ddd9-a359-4e97-b9d2-7800e5d8a1dd?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [93ad913106b6f0f6ec98bc2cfa788ff6d8085bd4](https://github.com/cockroachdb/cockroach/commits/93ad913106b6f0f6ec98bc2cfa788ff6d8085bd4):


```
=== RUN   TestCopyFromRandom
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCopyFromRandom4040902321
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
I240518 10:01:08.103567 1 (gostd) rand.go:250  [-] 1  random seed: -6111172529673371815
    copy_in_test.go:256: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Tenant/10/Table/106/1/960/0): "sql txn" meta={id=1906011f key=/Tenant/10/Table/106/1/0/0 iso=Serializable pri=0.00899420 epo=0 ts=1716026478.302179162,2 min=1716026477.421379670,0 seq=1001} lock=true stat=PENDING rts=1716026478.302179162,2 wto=false gul=1716026477.921379670,0
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestCopyFromRandom4040902321
--- FAIL: TestCopyFromRandom (14.38s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCopyFromRandom.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/copy/copy_test.TestCopyFromRandom [failed](https://tanzanite.cluster.engflow.com/invocations/default/1ce408f4-c13b-46c4-ae87-8bab761072c7?testReportRun=2&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [0ea8b622c5354a891cfb867d8d872a0746ba77b3](https://github.com/cockroachdb/cockroach/commits/0ea8b622c5354a891cfb867d8d872a0746ba77b3):


```
=== RUN   TestCopyFromRandom
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCopyFromRandom280335062
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
I240527 10:23:03.559772 1 (gostd) rand.go:250  [-] 1  random seed: 258427910704252356
    copy_in_test.go:256: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Tenant/10/Table/106/1/664/0): "sql txn" meta={id=b03a76ef key=/Tenant/10/Table/106/1/0/0 iso=Serializable pri=0.01700735 epo=0 ts=1716805396.009330544,2 min=1716805394.030776035,0 seq=1001} lock=true stat=PENDING rts=1716805396.009330544,2 wto=false gul=1716805394.530776035,0
    panic.go:626: -- test log scope end --
--- FAIL: TestCopyFromRandom (16.64s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCopyFromRandom.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/copy/copy_test.TestCopyFromRandom [failed](https://tanzanite.cluster.engflow.com/invocations/default/10238c9c-88eb-4860-bf34-e290a244c604?testReportRun=2&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [d102f9db95fb91646a980334207894831c079a79](https://github.com/cockroachdb/cockroach/commits/d102f9db95fb91646a980334207894831c079a79):


```
=== RUN   TestCopyFromRandom
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCopyFromRandom2717324991
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
I240612 10:19:11.826138 1 (gostd) rand.go:250  [-] 1  random seed: -9119351700112553178
    copy_in_test.go:256: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Tenant/10/Table/106/1/960/0): "sql txn" meta={id=8be4a1bc key=/Tenant/10/Table/106/1/0/0 iso=Serializable pri=0.03154669 epo=0 ts=1718187563.588373097,2 min=1718187562.142033364,0 seq=1001} lock=true stat=PENDING rts=1718187563.588373097,2 wto=false gul=1718187562.642033364,0
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestCopyFromRandom2717324991
--- FAIL: TestCopyFromRandom (15.83s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCopyFromRandom.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/copy/BUILD.bazel
+++ b/pkg/sql/copy/BUILD.bazel
@@ -15,6 +15,7 @@ go_test(
         "//pkg/base",
         "//pkg/cli/clisqlclient",
         "//pkg/kv",
+        "//pkg/kv/kvclient/kvcoord",
         "//pkg/kv/kvpb",
         "//pkg/kv/kvserver/kvserverbase",
         "//pkg/security/securityassets",

--- a/pkg/sql/copy/copy_in_test.go
+++ b/pkg/sql/copy/copy_in_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/cli/clisqlclient"
 	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/sql"
@@ -152,8 +153,13 @@ func TestCopyFromRandom(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	ctx := context.Background()
 	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
-	defer s.Stopper().Stop(context.Background())
+	defer s.Stopper().Stop(ctx)
+
+	// Disable pipelining. Without this, pipelined writes performed as part of the
+	// COPY can be lost, which can then cause the COPY to fail.
+	kvcoord.PipelinedWritesEnabled.Override(ctx, &s.ClusterSettings().SV, false)
 
 	if _, err := db.Exec(`
 		CREATE DATABASE d;

--- a/pkg/sql/copy_from.go
+++ b/pkg/sql/copy_from.go
@@ -1143,19 +1143,33 @@ func (c *copyMachine) insertRows(ctx context.Context, finalBatch bool) error {
 			// for the next batch.
 			return c.doneWithRows(ctx)
 		} else {
-			// It is currently only safe to retry if we are not in atomic copy
-			// mode & we are in an implicit transaction.
-			// NOTE: we cannot re-use the connExecutor retry scheme here as COPY
-			// consumes directly from the read buffer, and the data would no
-			// longer be available during the retry.
-			if c.implicitTxn && !c.p.SessionData().CopyFromAtomicEnabled && c.p.SessionData().CopyFromRetriesEnabled && errIsRetriable(err) {
-				log.SqlExec.Infof(ctx, "%s failed on attempt %d and is retrying, error %+v", c.copyFromAST.String(), r.CurrentAttempt(), err)
-				if c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry != nil {
-					if err := c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry(); err != nil {
-						return err
+			if errIsRetriable(err) {
+				log.SqlExec.Infof(ctx, "%s failed on attempt %d and with retriable error %+v", c.copyFromAST.String(), r.CurrentAttempt(), err)
+				// It is currently only safe to retry if we are not in atomic copy
+				// mode & we are in an implicit transaction.
+				//
+				// NOTE: we cannot re-use the connExecutor retry scheme here as COPY
+				// consumes directly from the read buffer, and the data would no
+				// longer be available during the retry.
+				if c.implicitTxn && !c.p.SessionData().CopyFromAtomicEnabled && c.p.SessionData().CopyFromRetriesEnabled {
+					log.SqlExec.Infof(ctx, "%s is retrying", c.copyFromAST.String())
+					if c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry != nil {
+						if err := c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry(); err != nil {
+							return err
+						}
 					}
+					continue
+				} else {
+					log.SqlExec.Infof(
+						ctx,
+						"%s is not retrying; "+
+							"implicit: %v; copy_from_atomic_enabled: %v; copy_from_retriable_enabled %v",
+						c.copyFromAST.String(), c.implicitTxn,
+						c.p.SessionData().CopyFromAtomicEnabled, c.p.SessionData().CopyFromRetriesEnabled,
+					)
 				}
-				continue
+			} else {
+				log.SqlExec.Infof(ctx, "%s failed on attempt %d and with non-retriable error %+v", c.copyFromAST.String(), r.CurrentAttempt(), err)
 			}
 			return err
 		}

